test/pytest/receiver/test_asapo_transfer.py

```diff
@@ -0,0 +1,277 @@
+@patch('asapo_transfer.create_asapo_transfer', create_mock_transfer)
+def test_run_transfer(worker, transfer_config_obj, file_list, hidra_metadata):
+
+    stop_event = Event()
+    x = threading.Thread(target=run_transfer, args=(worker, transfer_config_obj, 1, stop_event))
+    x.start()
+    sleep(2)
+    stop_event.set()
+    sleep(1)
+    x.join()
+    worker.send_message.assert_called_with(None, None)
+
+
+@patch('asapo_transfer.create_query', create_autospec(create_query, instance=True))
+@patch('asapo_transfer.create_asapo_transfer', create_mock_failing_transfer)
+def test_stop_error(worker, transfer_config_obj, file_list, hidra_metadata):
+
+    stop_event = Event()
+    x = threading.Thread(target=run_transfer, args=(worker, transfer_config_obj, 1, stop_event))
+    x.start()
+    sleep(5)
+    stop_event.set()
+    sleep(1)
+    x.join()
+    worker.send_message.assert_called_with(None, None)
+    # asapo_transfer is crashed and restarted several times
+    assert(len(worker.send_message.mock_calls) > 3)
+
+
+def test_path_parsing(asapo_transfer, file_list, hidra_metadata):
+    asapo_transfer.stop_run.is_set = Mock()
+    asapo_transfer.stop_run.is_set.side_effect = [False, False, True]
+    asapo_transfer.run()
+
+    asapo_transfer.asapo_worker.send_message.assert_called_with(file_list[0], hidra_metadata[0])
+
+    stream_list = ['foo1/bar1', 'foo1/bar2', 'foo2/bar1', 'foo2/bar2']*3
+    # ToDo: Refactor: Extract free function or so
+    for i, file_path in enumerate(file_list):
+        data_source, stream, file_idx = asapo_transfer.asapo_worker._parse_file_name(file_path)
+        assert file_idx == 1
+        assert stream == stream_list[i]
+        assert data_source == 'test001'
+
+
+def test_path_parse_no_offset(asapo_transfer):
+    # Offset is not encoded in the file path
+    file_path = "current/raw/data/tilt_27_048_041_data_070421.h5"
+    data_source, stream, file_idx = asapo_transfer.asapo_worker._parse_file_name(file_path)
+    assert file_idx == 70421
+
+
+def test_path_parse_nonzero_offset(asapo_transfer):
+    ragexp_template = "current/raw/(?P<scan_id>.*)_(?P<file_idx_offset>.*)_(?P<file_idx_in_scan>.*).h5"
+    asapo_transfer.asapo_worker.file_regex = ragexp_template
+
+    # Offset is encoded in the file path
+    file_path = "current/raw/data/tilt_27_048_041_data_070421_000001.h5"
+    data_source, stream, file_idx = asapo_transfer.asapo_worker._parse_file_name(file_path)
+    assert file_idx == 70422
+
+
+def test_path_parce_failed(asapo_transfer):
+    ragexp_template = "current/raw/(?P<scan_id>.*)_(?P<file_idx_offset>.*)_(?P<file_idx_in_scan>.*).h5"
+    asapo_transfer.asapo_worker.file_regex = ragexp_template
+    with pytest.raises(ValueError):
+        file_path = "current/raw/data/tilt_27_048_041_data_070421_foo.h5"
+        data_source, stream, file_idx = asapo_transfer.asapo_worker._parse_file_name(file_path)
+
+
+def test_init_query(asapo_transfer, file_list, hidra_metadata):
```

Comment:
What does this test test? It calls `run` twice on the same `asapo_transfer` object, which shouldn't happen in the real code? Actually, calling `run` twice should probably raise an exception.